### ISSUE TITLE
Scope in storage_dir on create

### ISSUE THREAD

**Issue by simonprev** [stavro/arc#10]
Would it be easy/straigthforward to have the scope in `storage_dir` after it passed throught some callbacks? My use case is: I have a model with a `slug` column and I want that column to be in the uploaded path. The content of this column is generated in a `before_insert` callback. The goal is to have a unique path for each record with an identifier persisted in the database. As it is right now in `arc`, the scope contains no `id` and no other fields set in callback.

Could it be possible to achieve this with `arc`? My hack right now is to inject the generated `slug` value in the controller before it goes in the changeset.

Thank you for this package, by the way, it’s really fun and easy to use :smile:

**Comment by stavro** [comment]
Let me rephrase your question to see if I understand:  You have some attributes on your model that you've changed through changesets, and you would like access to those changes inside of the upload definition module.

If you can share any of your model code to highlight the issue I would appreciate it.

Thanks!

**Comment by simonprev** [comment]
I have an attribute on my model that is changed through a callback

```elixir
before_insert :put_friendly_id, [:slug, 10]

def put_friendly_id(changeset, key, length) do
  Ecto.Changeset.put_change(changeset, key, generate_slug(length))
end

defp generate_slug(length) do
  :crypto.strong_rand_bytes(length) |> Base.url_encode64 |> binary_part(0, length)
end
```

The `slug` attributes is set _after_ the `cast_attachements` call in the changeset function.

Should I set the `slug` in the changeset function before the `cast_attachements`? (only on insert)

```elixir
def changeset(model, params \\ :empty) do
  model
  |> put_change(:slug, generate_slug(10))
  |> cast(params, @required_fields, @optional_fields)
  |> cast_attachments(params, ~w(media), ~w())
end
```

Thanks for the quick response!

**Comment by stavro** [comment]
> Should I set the slug in the changeset function before the cast_attachements? (only on insert)

I believe your second option is the only way it could possibly work.  Since as you've noted - the `slug` attribute is set after the call to cast_attachments when using `before_insert`.

However, let me make a quick update to `arc_ecto` which will enable this second option to work.  I'm currently only passing the root model (without changes applied) to the upload definition, but I've been wanting to apply the changes first, which I believe makes more sense.

**Comment by simonprev** [comment]
Alright, I also believe that the second option is the cleanest, I only thought of it when writing my comment.

Thanks for the update in `arc_ecto`, my migration from the `carrierwave` ruby gem is really easy (I’m rewriting a rails app to a phoenix app)

**Comment by stavro** [comment]
I'm glad to hear it!  :)

If there are any features you're leaving behind that you miss please let me know.

**Comment by stavro** [comment]
Can you try `arc_ecto` version `0.2.0` with the second option above and let me know the results? 

Thanks!

**Comment by stavro** [comment]
By the way - Just nitpicking, but `put_change` only accepts a changeset as the first argument, not a model.  So you'll need to switch the order (`cast`, then `put_change`, then `cast_attachments`).

```
  model
  |> cast(params, @required_fields, @optional_fields)
  |> put_change(:slug, generate_slug(10))
  |> cast_attachments(params, ~w(media), ~w())
```

**Comment by simonprev** [comment]
Yep it works as expected, thanks for the support :+1:
